### DISCUSSION TITLE
Docker Compose is now in the Docker CLI, try `docker compose up`

### DISCUSSION THREAD

**Discussion by AlexZeitler** [PDMLab/docker-compose#151] [General]
Well, good timing to [think about a 1.0.0](https://github.com/PDMLab/docker-compose/discussions/146) version 🤔

<img width="529" alt="image" src="https://user-images.githubusercontent.com/287480/117517841-a6e9d280-af9d-11eb-9141-1d7a20c8b7ea.png">

Some details: 

* https://github.com/docker/compose-cli/issues/1149
* https://github.com/docker/compose-cli#local-docker-compose

I guess, other  - already existing - libs, targeting plain `docker` command, will fill this space now.

**Comment by StefanScherer** [comment]
Yes, it's time to think about the new `docker compose` command that is built-in to the Docker CLI. There will be a phase where both are installed in parallel in eg. Docker Desktop. We at Docker try to migrate the users over by switching `docker-compose` commands to `docker compose` in a canary rollout to get more feedback of most common issues we should fix before we start removing the Python based binary in the future.

For this project it is a good chance to switch from `docker-compose` to `docker compose` independently or let the user choose to they don't need to install `docker-compose`, but only the new compose-cli (it's automatically installed with latest versions of Docker Desktop for Mac and Windows, but currently needs a manual installation for Linux)

- Docs: https://docs.docker.com/compose/cli-command/

Feedback in https://github.com/docker/compose-cli/issues is welcome if you find issues.

**Reply by Steveb-p** [reply]
@StefanScherer won't the functionality of `docker compose` be present in `dockerode` (i.e. Docker API) itself? This would in practice mean that this library becomes obsolete, since direct docker usage imo should be preferered over calling it via shell?

:+1: / :-1: ?

**Reply by StefanScherer** [reply]
Docker Compose is client side and not part of the Docker Engine API that eg. dockerode uses. I don‘t think that dockerode will obsolete this package soon.

**Reply by Steveb-p** [reply]
We'll need to see what changes there are between `docker-compose` and `docker compose`. I've seen that in a few cases json output can be produced, and that could be helpful with a few commands, like `ps` or `port`.

**Reply by AlexZeitler** [reply]
@StefanScherer I just started [the implementation](https://github.com/PDMLab/docker-compose/pull/164) for `docker compose`. I tried to install the new CLI for Linux as described [here](https://docs.docker.com/compose/cli-command/).

I restarted my VM, however calling `docker compose` results in:

```bash
docker compose
docker: 'compose' is not a docker command.
See 'docker --help'
```

Do you have any idea what could have gone wrong?

**Reply by StefanScherer** [reply]
@AlexZeitler Indeed, the `install_linux.sh` script hasn't been updated yet to install the `docker compose` CLI plugin that is used for the local `docker context`.

The `install_linux.sh` script currently only installs the Compose CLI 1.0.x as a replacement for the `docker` binary which is renamed to `com.docker.cli`. That is the initial work to support remote container contexts in AWS (ECS) and Azure (ACI) where a typical docker CLI plugin cannot handle all bits. That's why we had to replace the classic Docker CLI to have more control for these remote contexts.

As your use case is for the local contexts you need to install the Compose v2 which is released as a pre-release on GitHub until we have a GA version 2.0.x. 

For your environment where you only need `docker compose ` for the local context you don't need to run the `install_linux.sh` script, but only these commands

```shell
$  mkdir ~/.docker/cli-plugins
$  curl -L -o ~/.docker/cli-plugins/docker-compose https://github.com/docker/compose-cli/releases/download/v2.0.0-beta.3/docker-compose-linux-amd64
$  chmod +x ~/.docker/cli-plugins/docker-compose
$  docker compose 
```

The whole compose transition from Python to Go will also affect the repos and where you get the binaries at some point. So that's why there only is a v2.0.0-beta.3 release at the moment. And maybe the Compose CLI plugin will land in the docker-ce-cli package itself, but for now it's too early to put it there.

**Reply by AlexZeitler** [reply]
@StefanScherer Thanks 🙌 <img width="1366" alt="image" src="https://user-images.githubusercontent.com/287480/121081788-7b951600-c7dd-11eb-9cb1-e7116ea94780.png">

**Comment by cristianrgreco** [comment]
Hi, is there any activity around supporting compose v2? I'm aware this was released GA a year ago now. I work on a library called testcontainers and supporting compose v2 is a feature more and more people are asking for.

I saw[ this comment saying that dockerode should support this,](https://github.com/PDMLab/docker-compose/discussions/151#discussioncomment-738163) and I see there is a library called dockerode-compose, but this already seems to have been abandoned.

**Reply by AlexZeitler** [reply]
@cristianrgreco I would like to continue the work on #228 

Biggest effort to me right now is to get a test environment where I can easily switch between installed docker / compose plugin versions to track the issues mentioned in #228.

**Reply by AlexZeitler** [reply]
I think I can provide a new version which supports v2 later this week.

**Reply by AlexZeitler** [reply]
Here's the first green build for v2: 

https://github.com/PDMLab/docker-compose/actions/runs/4792472092

Known issues so far: 

* As mentioned in https://github.com/PDMLab/docker-compose/pull/228#issuecomment-1422895821, `compose` sometimes seems to return the exit code before it actually finished its work. Hacky workaround without introducing 3rd party libraries like `dockerode`: `setTimeout(500)`. This is far from being perfect but I want to get this thing out finally. I think not having v2 support at all becomes more of an issue than maybe having an issue that can be mitigated. Thoughts @cristianrgreco @Steveb-p?
* I've replaced `jest` by `vitest` because I really struggle to use jest since forever.
* v1 tests need to be moved over from jest to vitest before the release

**Reply by AlexZeitler** [reply]
I've released a preview version which includes v1 and v2 support: https://www.npmjs.com/package/docker-compose/v/0.23.20-preview.0

**Reply by AlexZeitler** [reply]
Import needs to be `import * as compose from "docker-compose/dist/v2";`

Will fix this for the release.

**Comment by AlexZeitler** [comment]
📦 0.24.0 is released with compose v2 support
